File: turtle_project_cpp_pkg/src/turtle_manager.cpp
#include <numbers>
#include <algorithm>
#include <random>

#include "rclcpp/rclcpp.hpp"
#include "turtlesim/srv/spawn.hpp"
#include "turtlesim/srv/kill.hpp"
#include "turtlesim/msg/pose.hpp"
#include "my_robot_interfaces/msg/turtle_loc_array.hpp"
#include "my_robot_interfaces/msg/turtle_loc.hpp"

using namespace std::placeholders;
using namespace std::chrono_literals;
 
class TurtleManagerNode : public rclcpp::Node
{
public:
    TurtleManagerNode() : Node("turtle_manager"), n_turtles(2)
    {
        this->declare_parameter("spawn_rate", 2.0);
        spawn_rate_ = static_cast<float>(this->get_parameter("spawn_rate").as_double());

        param_callback_handle_ = this->add_post_set_parameters_callback(
            std::bind(&TurtleManagerNode::parametersCallback, this, _1)
        );

        spawner_client_ = this->create_client<turtlesim::srv::Spawn>("spawn");
        killer_client_ = this->create_client<turtlesim::srv::Kill>("kill");
        pose1_sub_ = this->create_subscription<turtlesim::msg::Pose>("turtle1/pose", 10,
            std::bind(&TurtleManagerNode::callbackTurtle1Pose, this, _1));
        target_pub_ = this->create_publisher<my_robot_interfaces::msg::TurtleLocArray>("target_turtles", 10);
        spawn_timer_ = this->create_wall_timer(std::chrono::duration<double>(spawn_rate_), std::bind(&TurtleManagerNode::spawnTurtle, this));

        RCLCPP_INFO(this->get_logger(), "Turtle manager node started");
    }
 
private:
    rclcpp::Client<turtlesim::srv::Spawn>::SharedPtr spawner_client_;
    rclcpp::Client<turtlesim::srv::Kill>::SharedPtr killer_client_;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr pose1_sub_;
    rclcpp::Publisher<my_robot_interfaces::msg::TurtleLocArray>::SharedPtr target_pub_;
    rclcpp::TimerBase::SharedPtr spawn_timer_;
    PostSetParametersCallbackHandle::SharedPtr param_callback_handle_;

    std::array<float, 2> turtle1_pos;
    int n_turtles;
    float spawn_rate_;
    
    std::mutex available_turtles_mutex_;
    my_robot_interfaces::msg::TurtleLocArray available_turtles;

    void parametersCallback(const std::vector<rclcpp::Parameter>& parameters)
    {
        for (const auto &param: parameters) {
            spawn_rate_ = static_cast<float>(param.as_double());
            spawn_timer_->cancel();
            spawn_timer_->reset();
            spawn_timer_ = this->create_wall_timer(std::chrono::duration<double>(spawn_rate_), std::bind(&TurtleManagerNode::spawnTurtle, this));
        }
    }

    void callbackTurtle1Pose(const turtlesim::msg::Pose::SharedPtr msg)
    {
        turtle1_pos = {msg->x, msg->y};

        std::string turtle_to_kill;

        {
            std::lock_guard<std::mutex> lock(available_turtles_mutex_);
            auto it = std::find_if(available_turtles.turtles.begin(),
                available_turtles.turtles.end(),
                [this](const my_robot_interfaces::msg::TurtleLoc& turtle) {
                    return std::abs(turtle.x - turtle1_pos[0]) <= 0.5 && std::abs(turtle.y - turtle1_pos[1]) <= 0.5;
                }
            );

            if (it != available_turtles.turtles.end()) {
                turtle_to_kill = it->name;
            }
        }

        if (!turtle_to_kill.empty()) {
            killTurtle(turtle_to_kill);
        }
    }

    void spawnTurtle()
    {
        while (!spawner_client_->wait_for_service(1s))
        {
            RCLCPP_WARN(this->get_logger(), "Waiting for the Turtlesim Spawn server ...");
        }

        std::mt19937_64 rng(std::random_device{}()); // Mersenne Twister engine
        std::uniform_real_distribution<double> dist(0.0, 1.0);

        auto request = std::make_shared<turtlesim::srv::Spawn::Request>();
        request->x = dist(rng) * 10;
        request->y = dist(rng) * 10;
        request->theta = (dist(rng) * 2 - 1) *std::numbers::pi;
        request->name = "turtle" + std::to_string(n_turtles);

        n_turtles++;

        spawner_client_->async_send_request(request,
            [this, request](rclcpp::Client<turtlesim::srv::Spawn>::SharedFuture future) {
                auto response = future.get();
                auto turtle = my_robot_interfaces::msg::TurtleLoc();
                turtle.name = response->name;
                turtle.x = request->x;
                turtle.y = request->y;
                available_turtles.turtles.push_back(turtle);
                target_pub_->publish(available_turtles);
            }
        );
    }

    void killTurtle(const std::string& name)
    {
        while (!killer_client_->wait_for_service(1s))
        {
            RCLCPP_WARN(this->get_logger(), "Waiting for the Turtlesim Kill server ...");
        }

        auto request = std::make_shared<turtlesim::srv::Kill::Request>();
        request->name = name;

        killer_client_->async_send_request(request,
            [this, request](rclcpp::Client<turtlesim::srv::Kill>::SharedFuture future) {
                auto response = future.get();
            }
        );

        removeTurtleByName(request->name);
        target_pub_->publish(available_turtles);
    }

    void removeTurtleByName(const std::string& name_to_remove)
    {
        std::lock_guard<std::mutex> lock(available_turtles_mutex_);
        auto it = std::find_if(available_turtles.turtles.begin(), available_turtles.turtles.end(),
        [&](const my_robot_interfaces::msg::TurtleLoc& turtle) {
            return turtle.name == name_to_remove;
        });

        if (it != available_turtles.turtles.end()) {
            available_turtles.turtles.erase(it);
        } else {
            RCLCPP_WARN(this->get_logger(), "Turtle not found: %s", name_to_remove.c_str());
        }
    }
};
 
int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TurtleManagerNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: turtle_project_py_pkg/turtle_project_py_pkg/turtle_manager.py
#!/usr/bin/env python3
from functools import partial

import rclpy
from rclpy.node import Node
from rclpy.parameter import Parameter
from turtlesim.srv import Spawn
from turtlesim.msg import Pose
from turtlesim.srv import Kill

import numpy as np 
 
class TurtleManagerNode(Node):
    def __init__(self):
        super().__init__("turtle_manager")
        self.declare_parameter("spawn_rate", 2.0)
        self.spawn_rate_ = self.get_parameter("spawn_rate").value
        self.add_post_set_parameters_callback(self.parameters_callback)

        self.spawned_turtles_ = []
        self.turtle_pose_subs_ = self.create_subscription(Pose, "turtle1/pose", self.callback_turtle_pose, 10)
        self.spawner_client = self.create_client(Spawn, "spawn")
        self.killer_client = self.create_client(Kill, "kill")
        self.get_logger().info("Turtle manager has been started")

    def parameters_callback(self, params: list[Parameter]):
        for param in params:
            if param.name == "spawn_rate":
                self.spawn_rate_ = param.value
                self.get_logger().info("Spawn rate has been changed to: " + str(self.spawn_rate_))
                self.spawn_timer_.cancel()
                self.start_spawning()
    
    ### Listening to turtle 1 pose to kill other turtle ###
    def callback_turtle_pose(self, msg: Pose):
        turtle_indices = [i for i, t in enumerate(self.spawned_turtles_) if 
            np.isclose(msg.x, t["x"], atol=0.5) and np.isclose(msg.y, t["y"], atol=0.5)]
        for i in turtle_indices:
            t = self.spawned_turtles_.pop(i)
            self.kill_turtle(t["name"])

    ### Killing turtles ###
    def kill_turtle(self, name: str):
        while not self.killer_client.wait_for_service(timeout_sec=1.0):
            self.get_logger().warn("Waiting for Killer server ...")

        request = Kill.Request()
        request.name = name

        future = self.killer_client.call_async(request)
        future.add_done_callback(partial(self.callback_kill_turtle, request=request))

    def callback_kill_turtle(self, future, request):
        try:
            response = future.result()
            # self.get_logger().info("Turtle %s killed" % request.name)
        except Exception as e:
            self.get_logger().error("Service call failed %r" % (e,))

    ### Spawning turtles ###
    def start_spawning(self):
        n = 0
        def encapsulated_spawn_turtle():
            nonlocal n
            x = np.random.random()*10
            y = np.random.random()*10
            theta = np.random.random()*2*np.pi
            name = "turtle_%d" % (n+2)
            self.spawn_turtle(x, y, theta, name)
            n += 1
        self.spawn_timer_ = self.create_timer(self.spawn_rate_, encapsulated_spawn_turtle)

    def spawn_turtle(self, x: float, y: float, theta: float, name: str):
        while not self.spawner_client.wait_for_service(timeout_sec=1.0):
            self.get_logger().warn("Waiting for Spawner server ...")

        request = Spawn.Request()
        request.x = x
        request.y = y
        request.theta = theta
        request.name = name

        future = self.spawner_client.call_async(request)
        future.add_done_callback(partial(self.callback_spawn_turtle, request=request))

    def callback_spawn_turtle(self, future, request):
        response = future.result()
        # self.get_logger().info("Turtle %s spawned at (%.2f, %.2f, %.2f)" % (response.name, request.x, request.y, request.theta))
        added_turtle = {"name": response.name, "x": request.x, "y": request.y, "theta": request.theta}
        self.spawned_turtles_.append(added_turtle)
 
 
def main(args=None):
    rclpy.init(args=args)
    node = TurtleManagerNode()
    node.start_spawning()
    rclpy.spin(node)
    rclpy.shutdown()
 
 
if __name__ == "__main__":
    main()